fix(nvidia-safety)!: correct NeMo Guardrails API endpoint

r-bit-rry opened the pull request:
Fixes #4189 - NVIDIA safety provider endpoint issues.

## Changes

Added `api_mode` config to support both NeMo Guardrails deployments:

- **`microservice`** (default): Enterprise NIM using `/v1/guardrail/checks`
- **`openai`**: Open-source toolkit using `/v1/guardrail/chat/completions`

### Additional improvements:
- Configurable `timeout` (default 60s)
- Configurable `blocked_message`
- Temperature validation (0-2)
- Better error handling with clear error messages
- Unknown API errors now raise instead of silently passing

## Configuration

```yaml
api_mode: microservice  # or 'openai' for open-source toolkit
timeout: 60
blocked_message: "I'm sorry, I can't respond to that."
```

## Testing

- 19 unit tests covering both API modes
- All 47 NVIDIA provider tests pass
- Pre-commit hooks pass

The referenced documentation:
https://docs.nvidia.com/nemo/guardrails/latest/user-guides/server-guide.html
https://docs.nvidia.com/nemo/microservices/25.8.0/api/guardrails.html

ashwinb commented:
<img width="644" height="243" alt="image" src="https://github.com/user-attachments/assets/2faeebb8-4496-4aa5-8170-970ee7f52aa2" />

where is the mentioned reproduction script? Either this could be included as a link to a Gist, or better still -- the PR summary should perhaps either be hand edited to look more like what a human engineer would write (just do not include menial details which are irrelevant for a change of such small magnitude) or the LLM could be guided with this precise instruction so the generated Summary feels like that.

r-bit-rry commented on src/llama_stack/providers/remote/safety/nvidia/config.py:
```diff
@@ -25,16 +33,39 @@ class NVIDIASafetyConfig(BaseModel):
+        description="Sampling temperature for the guardrails model, between 0 and 2",
+    )
+    timeout: int = Field(
+        default=60,
```
had to be added due to the move from requests to httpx

leseb commented on src/llama_stack/providers/remote/safety/nvidia/nvidia.py:
```diff
@@ -69,8 +73,15 @@ async def run_shield(
+        shield_params = shield.params or {}
+        model = shield_params.get("model") or shield.provider_resource_id
+        if not model:
+            raise ValueError(
```
this should raise ModelNotFoundError instead?

r-bit-rry replied:
The suggestion to use ModelNotFoundError makes sense at first glance, but here's why it doesn't quite fit:

  What ModelNotFoundError is designed for:
  ```
  # User asks for a model that doesn't exist in the system
  client.models.get("llama-doesnt-exist")  # → ModelNotFoundError("llama-doesnt-exist")
  # Message: "Model 'llama-doesnt-exist' not found. Use 'client.models.list()' to list available Models."
```
  What's actually happening in our code:
  ```
  # Shield exists, but its 'model' field is empty/None
  shield.params.get("model")  # → None
  shield.provider_resource_id  # → None
  # The shield was registered without specifying which model to use
```
  If we used ModelNotFoundError(model) where model is None or empty string, we'd get a confusing message like:
  "Model '' not found. Use 'client.models.list()' to list available Models."

 listing models won't help because the problem is the shield's config, not a missing model resource.

  I find that the current ValueError message is more helpful:
  "Shield {shield_id} does not have a model configured. Set 'model' in params or provider_resource_id."

  This tells the user exactly what's wrong and how to fix it.
  As for provider_resource_id, the shield configuration is user-provided when they call register_shield():
```
  client.shields.register(
      shield_id="my-guard",
      provider_shield_id="meta/llama-3.2-3b-instruct",  # → becomes provider_resource_id
      provider_id="nvidia",
      params={"model": "override-model"}  # optional override
  )
```
  So if provider_resource_id and params["model"] are both empty, it means the user registered the shield without specifying which model to use.
The main difference:
The provider_resource_id for NVIDIA safety isn't a llama-stack Model resource, it's just a string identifier that gets passed to the NeMo Guardrails API (like "meta/llama-3.2-3b-instruct"). We're not doing a lookup in llama-stack's model registry.

leseb commented on src/llama_stack/providers/remote/safety/nvidia/nvidia.py:
```diff
@@ -84,83 +95,167 @@ class NeMoGuardrails:
+    async def _guardrails_post(self, path: str, data: Any | None) -> dict[str, Any]:
+        """Make a POST request to the guardrails service."""
+        headers = {"Accept": "application/json", "Content-Type": "application/json"}
+        url = f"{self.guardrails_service_url}{path}"
```
use urllib instead of string concat?

leseb commented on src/llama_stack/providers/remote/safety/nvidia/nvidia.py:
```diff
@@ -84,83 +95,167 @@ class NeMoGuardrails:
+            logger.error(
+                f"Guardrails service error: type={error_type}, code={error_code}, message={error.get('message', 'unknown')}"
+            )
+            raise RuntimeError(f"Failed to run guardrails check: {error.get('message', 'Unknown error')}")
```
this should be reworked to its own function + type error model i guess, this is adding too much to this function

leseb commented on src/llama_stack/providers/remote/safety/nvidia/nvidia.py:
```diff
@@ -84,83 +95,167 @@ class NeMoGuardrails:
+            raise RuntimeError(f"Failed to run guardrails check: {error.get('message', 'Unknown error')}")
+
+        # Check for legacy format with status field
+        if response.get("status") == "blocked":
```
same for this block

r-bit-rry replied:
Are you sure, its just 3 lines effectively returning response?

leseb commented on .gitignore:
```diff
@@ -37,3 +37,4 @@ docs/docs/api-experimental/
 docs/docs/api/
 tests/integration/client-typescript/node_modules/
 .ts-client-checkout/
+.beads
```
why?

leseb reviewed (changes requested):
I'm not super knowledgeable on NeMo but as far as i can tell this PR underwent multiple reviews rounds by people with the right expertise. Someone from NVIDIA highlighted that the current PR was fixing a bug too. It seems to me that the testing has been pretty thorough so let's proceed with this one.
Nice work and sorry for the lack of attention.

Also, where is the breaking change in this PR?

r-bit-rry commented:
@m-misiura thanks for your review:
1. I Renamed API modes: MICROSERVICE → GUARDRAIL_CHECKS, OPENAI → GUARDRAIL_CHAT_COMPLETIONS in config.py. Added a field_validator that maps old string values ("microservice", "openai") to the new enum values for backward compat with existing configs.

2. guardrails_data parsing for /v1/guardrail/checks: _run_guardrail_checks in nvidia.py now parses guardrails_data.log.activated_rails and checks for any rail with stop: true as the primary detection signal. Blocking rail names are included in the violation metadata. The legacy status == "blocked" check is kept as a fallback for older NIM versions that may not return guardrails_data.

3. Hybrid endpoint in safety provider: no changes here. As @leseb noted, this was reviewed and accepted. /v1/guardrail/checks is the endorsed NIM endpoint for guardrail evaluation, and stripping it would break integration with the closed-source NeMo microservice.

Code was rebased with upstream main
